tests/utils/common/common.py

```diff
@@ -32,6 +28,79 @@
 import traceback
 
 
+class Result:
+    def __init__(self, stdout, stderr, exited):
+        self.stdout = stdout
+        self.stderr = stderr
+        self.exited = exited
+        self.return_code = exited
+
+
+class Connection:
+    def __init__(self, host, user, port, connect_timeout, connect_kwargs={}):
+        self.host = host
+        self.user = user
+        self.port = port
+        self.connect_timeout = connect_timeout
+        self.connect_kwargs = connect_kwargs
+
+        self.key_filename = None
+
+        for k in connect_kwargs.keys():
+            if k == "key_filename":
+                self.key_filename = connect_kwargs[k]
+            else:
+                raise NotImplementedError(f"Argument {k} is not implemented")
+
+    def get_connect_args(self):
+        if self.key_filename is not None:
+            key_arg = ["-i", self.key_filename]
+        else:
+            key_arg = []
+
+        args = (
+            ["ssh"]
+            + key_arg
+            + [
+                "-p",
+                str(self.port),
+                "-o",
+                f"ConnectTimeout={self.connect_timeout}",
+                "-o",
+                "UserKnownHostsFile=/dev/null",
+                "-o",
+                "StrictHostKeyChecking=no",
+                f"{self.user}@{self.host}",
+            ]
+        )
+
+        return args
+
+    def run(self, command, warn=False, hide=False):
+        ssh_command = self.get_connect_args() + [command]
+
+        proc = subprocess.Popen(
+            ssh_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        stdout, stderr = proc.communicate()
+        stdout, stderr = (stdout.decode(), stderr.decode())
+        if not hide:
+            print(stdout)
+            print(stderr)
+        proc.wait()
```

Comment:
I'm aware that this could be simplified by using `subprocess.run()` instead. But the `capture_output` argument is a pretty recent addition which doesn't work on my distro yet.

Reply:
`capture_output` seems to be supported since 3.7.0 => https://github.com/python/cpython/blob/v3.7.0/Lib/subprocess.py#L446 aka >4 years ago...
if you can not use this I have to assume you are running 3.6 or older, which is EoL for a while: https://endoflife.date/python
You should consider updating your distro or at least your python installation and then simplify this to use `subprocess.run()`.

Reply:
I will have to side with @mzedel here

Reply:
I've been dreading this Ubuntu upgrade for a while (and Krzysztof's experience didn't help...), but it probably is time. However, this requires some planning as I need to do a full backup before attempting this, and I know from experience it will take several hours. I'll keep this PR on a feature branch for now, and I'll try to schedule some time next sprint.

Note: Python may have EOL'ed it, but Linux distributions are still maintaining it. Debian is even on a version prior to mine, and it's still actively patched.

Reply:
Yes, but isn't a fair solution here, to go with what is supported in the build-host base OS version?

Reply:
Then you can smudge it locally in the meantime?

Reply:
As a general rule I think some compatibility code should be allowed. There are good reasons to stay on stable distros for some time.

But in this case, you guys are right. My Ubuntu is getting very old, and is 9/10 into its lifespan. I really should upgrade. It's not like I don't want to get it done, but it's like doing the dishes you know! 🙂

This is just one piece of a much bigger puzzle, so there is no urgency to merge this.

Reply:
Alright, after upgrading I finally got the chance to fix this, and actually test it.

However, I had to change it a bit, because it needs to use `Popen` when the command is supposed to be in the background, which some tests need, so I added an argument specifically for this case. @lluiscampos, @oleorhagen, @mzedel: I would appreciate another look at the section around [here](https://github.com/mendersoftware/mender-image-tests/pull/71/files#diff-6f14d16e63d4ed9375105c4f121b9c05bde36d5c3a74ca0387409152ac1c481dR89).

Reply:
lgtm :)